File: scala/yacht/src/main/scala/Yacht.scala
object Yacht {

  def score(dice: List[Int], category: String): Int =
    category match {
      case "ones" => scoreSingle(dice, 1)
      case "twos" => scoreSingle(dice, 2)
      case "threes" => scoreSingle(dice, 3)
      case "fours" => scoreSingle(dice, 4)
      case "fives" => scoreSingle(dice, 5)
      case "sixes" => scoreSingle(dice, 6)
      case "full house" => scoreFullHouse(dice)
      case "four of a kind" => scoreFourOfAKind(dice)
      case "little straight" => scoreLittleStraight(dice)
      case "big straight" => scoreBigStraight(dice)
      case "yacht" => scoreYacht(dice)
      case "choice" => scoreChoice(dice)
    }

  private def scoreSingle(dices: List[Int], dice: Int) : Int = dices.filter(_ == dice).sum

  private def scoreFullHouse(dice: List[Int]): Int =
    dice.groupBy(identity).view.mapValues(_.size).toList.sortBy(_._2) match {
      case List((_, 2), (_, 3)) => dice.sum
      case _ => 0
    }

  private def scoreFourOfAKind(dice: List[Int]): Int =
    dice.groupBy(identity).view.mapValues(_.size).toList.sortBy(_._2) match {
      case List((_, 1), (die, 4)) => die * 4
      case List((die, 5)) => die * 4
      case _ => 0
    }

  private def scoreLittleStraight(dice: List[Int]): Int =
    dice.sorted match {
      case List(1, 2, 3, 4, 5) => 30
      case _ => 0
    }

  private def scoreBigStraight(dice: List[Int]): Int =
    dice.sorted match {
      case List(2, 3, 4, 5, 6) => 30
      case _ => 0
    }

  private def scoreYacht(dice: List[Int]): Int = if (dice.distinct.size == 1) 50 else 0
  private def scoreChoice(dices: List[Int]): Int = dices.sum
}

File: clojure/yacht/src/yacht.clj
(ns yacht)

(defn score-single [scores single]
  (apply + (filter #(= single %) scores)))

(defn score-full-house [scores]
  (let [counts (sort-by second (into (vector) (frequencies scores)))]
    (if (= [2 3] (map second counts)) (apply + scores) 0)))

(defn score-four-of-a-kind [scores]
  (let [counts (sort-by second (into (vector) (frequencies scores)))]
    (case (map second counts)
      [1 4] (* 4 (first (second counts)))
      [5] (* 4 (first scores))
      0)))

(defn score-little-straight [scores]
  (if (= [1 2 3 4 5] (sort scores)) 30 0))

(defn score-big-straight [scores]
  (if (= [2 3 4 5 6] (sort scores)) 30 0))

(defn score-choice [scores]
  (apply + scores))

(defn score-yacht [scores]
  (if (= 1 (count (distinct scores))) 50 0))

(defn score [scores category]
    (case category
      "ones" (score-single scores 1)
      "twos" (score-single scores 2)
      "threes" (score-single scores 3)
      "fours" (score-single scores 4)
      "fives" (score-single scores 5)
      "sixes" (score-single scores 6)
      "full house" (score-full-house scores)
      "four of a kind" (score-four-of-a-kind scores)
      "little straight" (score-little-straight scores)
      "big straight" (score-big-straight scores)
      "choice" (score-choice scores)
      "yacht" (score-yacht scores)))
